pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go

```diff
@@ -0,0 +1,465 @@
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting EndpointSliceMirroring controller")
+	defer klog.Infof("Shutting down EndpointSliceMirroring controller")
+
+	if !cache.WaitForNamedCacheSync("endpoint_slice_mirroring", stopCh, c.endpointsSynced, c.endpointSlicesSynced, c.servicesSynced) {
+		return
+	}
+
+	klog.V(2).Infof("Waiting for application of %s label", discovery.LabelSkipMirror)
+	c.waitForSkipLabelApplication()
+	klog.V(2).Infof("%s label successfully applied", discovery.LabelSkipMirror)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
+
+// This controller should not run until the EndpointSlice skip mirroring label
+// has been applied to Endpoints managed by the Endpoints controller. This
+// function waits until that's true.
+func (c *Controller) waitForSkipLabelApplication() error {
+	return wait.PollImmediate(2*time.Second, 5*time.Minute, func() (bool, error) {
+		endpoints, err := c.endpointsLister.List(labels.Everything())
+		if err != nil {
+			return false, err
+		}
+		services, err := c.serviceLister.List(labels.Everything())
+		if err != nil {
+			return false, err
+		}
+
+		servicesByName := map[string]*v1.Service{}
+		for _, svc := range services {
+			servicesByName[svc.Name] = svc
+		}
+
+		numEndpointsWithProblems := 0
+		for _, ep := range endpoints {
+			svc, ok := servicesByName[ep.Name]
+			if ok && svc.Spec.Selector != nil {
+				labelVal, ok := ep.Labels[discovery.LabelSkipMirror]
+				if !ok || labelVal != "true" {
+					klog.V(4).Infof("Expected %s Endpoints resource to have %s label set to true, got %s", ep.Name, discovery.LabelSkipMirror, labelVal)
+					numEndpointsWithProblems++
+				}
+			}
+		}
+
+		if numEndpointsWithProblems > 0 {
+			klog.V(2).Infof("Waiting to start EndpointSliceMirroring controller until Endpoints controller applies %s label (waiting for %d/%d Endpoints)", discovery.LabelSkipMirror, numEndpointsWithProblems, len(endpoints))
+			return false, nil
+		}
+
+		return true, nil
+	})
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same service
+// at the same time
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncEndpoints(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		klog.Warningf("Error mirroring EndpointSlices for %q Endpoints, retrying. Error: %v", key, err)
```

Comment:
I've moved to this approach because it reads more naturally to me to have the name followed by the type ("example for foo/bar Endpoints" vs "example for Endpoints foo/bar". I think it's just that I think the name is more important than the type here, or maybe it's the weird pluralization of Endpoints? Not quite sure why I like having the name first, but I'm open to changing it.